### PR TITLE
make winning possible

### PR DESCRIPTION
**Pull request by DerpyNewbie**

### CODE REVIEW

File: Source/TicTacToe/TicTacToeBlockGrid.h [hunk 1]
```diff
@@ -26,9 +27,15 @@ class ATicTacToeBlockGrid : public AActor
 	/** How many blocks have been clicked */
 	int32 Score;
 
-	/** Current turn role by int8. 1 to circle, -1 to cross.*/
+	/** Current turn role by int. 1 to circle, -1 to cross. */
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlockGrid.h]: test

File: Source/TicTacToe/TicTacToeBlock.cpp [hunk 2]
```diff
@@ -105,7 +98,19 @@ void ATicTacToeBlock::Highlight(bool bOn)
 		BlockMesh->SetMaterial(0, HighlightMaterial);
 	}
 	else
-	{
-		BlockMesh->SetMaterial(0, BaseMaterial);
+	{	// Using GetMaterialByType for fail safe
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlock.cpp]: Unclear what you mean here.

File: Source/TicTacToe/TicTacToeBlock.cpp [hunk 3]
```diff
@@ -105,7 +98,19 @@ void ATicTacToeBlock::Highlight(bool bOn)
...
+{
+	UMaterialInterface* matI = GetMaterialByType(_type);
+	BlockMesh->SetMaterial(0, matI);
+}
+
+void ATicTacToeBlock::SetType(int _type)
+{
+	SetMaterialByType(_type);
+	BlockType = _type;
+}
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlock.cpp:136]: Why do we need SetMaterialByType, just do it like BlockMesh->SetMaterial(0, GetMaterialByType(BlockType)); above.
Or use a method for every assignment. You could just do SetType(BlockType) above instead.

Either way, you are repeating yourself here.

File: Source/TicTacToe/TicTacToeBlockGrid.cpp [hunk 4]
```diff
@@ -74,4 +77,127 @@ void ATicTacToeBlockGrid::ChangeTurn()
...
+		CurrentGameState = winState;
+		HandleGameEnd();
+	}
+	else
+	{
+		AddScore();
+		ChangeTurn();
+		OnNextTurn();
+	}
+}
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlockGrid.cpp:110]: Remove executedBlock from this and move the logic that references it into the actual Block. Everything else you can leave because it doesn't need the block.

You could just use the Grid reference inside the Block for all of this too.

File: Source/TicTacToe/TicTacToeBlockGrid.cpp [hunk 5]
```diff
@@ -74,4 +77,127 @@ void ATicTacToeBlockGrid::ChangeTurn()
...
+
+int ATicTacToeBlockGrid::GetPosition(int num_x, int num_y)
+{
+	return  num_x + (num_y * Size);
+}
+
+int ATicTacToeBlockGrid::GetPosition(ATicTacToeBlock* block_pos)
+{
+	return BlockArray.Find(block_pos);
+}
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlockGrid.cpp:177]: Rename these into GetIndex. Position is confusing.

**Review by I-K0haku-I — CHANGES_REQUESTED** [review]

File: Source/TicTacToe/TicTacToeBlock.cpp (hunk 3 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlock.cpp:136]: I couldn't figure out how to set materials on `TicTacToeBlockGrid`. 
(i tried to do `executedBlock->BlockMesh->SetMaterial()` but didn't work at least on my visual studio intellisense.)

So, i made `SetMaterialByType()` to set material on `TicTacToeBlockGrid`.
but i actually didn't like calling setter 2 times in `TicTacToeBlockGrid`.

so i decided to add `SetType()` to set type with one method.
since it's not used by `TicTacToeBlockGrid`, `SetMaterialByType()` should be private/protected method or deleted.

TL:DR
`SetMaterialByType()` should be deprecated or private/protected method, since i made `SetType()` for setting types.

File: Source/TicTacToe/TicTacToeBlock.cpp (hunk 2 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlock.cpp]: Setting `BaseMaterial` directly made "confused block" like this, when i failed to fix "if returning check" on the `line:91`.
`(BlockType = 1 (Circle), Material = BaseMaterial (should be OrangeMaterial))`

So, i decided to instead of changing to `BaseMaterial` directly, changing from `GetMaterialByType()` by current `BlockType` to make sure the unhighlighted blocks are having corrected materials.

TL:DR
I changed BaseMaterial to GetMaterialByType so when accidentally called this method, it won't reset materials. (that's what i mean by fail safe)

Comment (I-K0haku-I) [reply Source/TicTacToe/TicTacToeBlock.cpp]: Unnecessary information then. Can be removed.

File: Source/TicTacToe/TicTacToeBlock.cpp (hunk 3 above)
Comment (I-K0haku-I) [reply Source/TicTacToe/TicTacToeBlock.cpp:136]: Create a ResetMaterial method and use SetType for actually setting it. SetType might not even be necessary if we move most of the calls into Block.

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 5 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp:177]: Renamed to `GetIndex()` on latest commit `5e633dc351346d17ab6b0cb65152d6ad49530768`

File: Source/TicTacToe/TicTacToeBlock.cpp (hunk 2 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlock.cpp]: Removed unnecessary comment on latest commit `5e633dc351346d17ab6b0cb65152d6ad49530768`

File: Source/TicTacToe/TicTacToeBlock.cpp (hunk 3 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlock.cpp:136]: Changed `SetMaterialByType()`'s access specifier to private on latest commit `5e633dc351346d17ab6b0cb65152d6ad49530768`
(still used in `ATicTacToeBlock` class. i feel like it's little bit more easy to read.)

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 4 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp:110]: I'm still not sure about this.
i think those stuff should be in `TicTacToeBlockGrid`.
since they are deciding what to do by stuff what `ATicTacToeBlockGrid` have.

File: Source/TicTacToe/TicTacToeBlockGrid.cpp [hunk 6]
```diff
@@ -74,4 +77,128 @@ void ATicTacToeBlockGrid::ChangeTurn()
 	CurrentType = CurrentType / -1;
 }
 
+void ATicTacToeBlockGrid::HandleTurn(ATicTacToeBlock* executedBlock)
+{
+	UE_LOG(LogTemp, Verbose, TEXT("Processing turn!"));
+
+	if (CurrentGameState != 0 || executedBlock->BlockType != 0) {
+		return;
```
Comment (DerpyNewbie) [inline Source/TicTacToe/TicTacToeBlockGrid.cpp]: Should be throwing exception since they were not able to handle turn?

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 4 above)
Comment (I-K0haku-I) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp:110]: @DerpyNewbie 
That's fine, you can keep it like that then. Probably just a taste thing.

But the main thing is how confusing the code latter part of the code is because you are doing so much at once. The worst offender is ```checkWin``` because it returns both block type and the game state even though you catch it as ```winState```. That's probably why the conditional blocks after that look so confusing.

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 6 above)
Comment (I-K0haku-I) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp]: I don't understand why you need an exception. If the turn is not possible then you can't click on it.

Also, you have this check in ```HandleClicked``` in the actual Block already.

File: Source/TicTacToe/TicTacToeBlock.cpp [hunk 7]
```diff
@@ -62,40 +65,35 @@ void ATicTacToeBlock::OnFingerPressedBlock(ETouchIndex::Type FingerIndex, UPrimi
...
 }
 
 void ATicTacToeBlock::Highlight(bool bOn)
 {
-	// Do not highlight if the block has already been activated.
-	if (BlockType != 0)
+	UE_LOG(LogTemp, VeryVerbose, TEXT("TILEARRAYPOS: %d, BLOCKTYPE: %s, INMATCH: %s, WINSTATE: %s"), OwningGrid->GetIndex(this), TOSTR(BlockType), TOSTR(OwningGrid->IsInMatch), TOSTR(OwningGrid->WinState));
+	// Do not highlight if the block has already been activated or if the current game state has not set to 0.
+	if (BlockType != EBlockType::None || !OwningGrid->IsInMatch)
 	{
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlock.cpp:98]: Could make a ```CanClick``` method since this appears twice now. Would be more clear what happens too.

**Review by I-K0haku-I — CHANGES_REQUESTED** [review]

File: Source/TicTacToe/TicTacToeBlockGrid.cpp [hunk 8]
```diff
@@ -64,14 +75,141 @@ void ATicTacToeBlockGrid::AddScore()
...
+	UE_LOG(LogTemp, Verbose, TEXT("Processing turn!"));
+
+	if (!IsInMatch || executedBlock->GetBlockType() != EBlockType::None) {
+		return;
+	}
+
+	// Set type / material
+	executedBlock->SetType(CurrentType);
+
+	WinState = CheckWin(executedBlock);
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlockGrid.cpp]: Rename to ```CheckWinState``` if you want to keep it like this.

**Review by I-K0haku-I — CHANGES_REQUESTED** [review]

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 8 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp]: Renamed to `CheckWinState()` on commit `1989be663cc69a2a645061378aae6769fd9c8533`

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 6 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp]: Removed if statement on commit `1989be663cc69a2a645061378aae6769fd9c8533`
(i thought it's useful if it throws exception so AI can notice they failed to click ;w;)

File: Source/TicTacToe/TicTacToeBlock.cpp (hunk 7 above)
Comment (DerpyNewbie) [reply Source/TicTacToe/TicTacToeBlock.cpp:98]: Added `IsClickable()` on commit `e09f9c72704772da6563dc98164db426c90aa862`

File: Source/TicTacToe/TicTacToeBlockGrid.cpp (hunk 4 above)
Comment (I-K0haku-I) [reply Source/TicTacToe/TicTacToeBlockGrid.cpp:110]: Another thing: The EWinState you receive is ONLY used in this HandleTurn method and nowhere else, correct me if I'm wrong.
You could call the method in the conditional block inside CheckWinState instead of setting the Enums and it would be the same.

Furthermore, I should move the 'can win' method from my AI to the static class. This way, we can just call it in CheckWinState. So no need to change anything right now. (Same with GetIndex)

File: Source/TicTacToe/TicTacToeBlockGrid.cpp [hunk 9]
```diff
@@ -126,78 +125,80 @@ int ATicTacToeBlockGrid::CheckWin(ATicTacToeBlock* checkBlock)
...
 
 void ATicTacToeBlockGrid::HandleGameEnd()
 {
-	UE_LOG(LogTemp, Log, TEXT("Ending!"));
+	UE_LOG(LogGrid, Log, TEXT("Ending!"));
+	// Disable player inputs
+	IsInMatch = false;
 
-	//TODO: Implement game end
 	// how do i implement showing reset button? :thinking:
```
Comment (I-K0haku-I) [inline Source/TicTacToe/TicTacToeBlockGrid.cpp:163]: Watch Tutorial for UI? 🤔
